File: Hiq.py
import sys
import time
from random import shuffle

############################################
#  Created by Vishal Kulkarni on 11/29/16.
#     VVK27 | CPL635 | vvk27@njit.edu
############################################


#################
##### BOARD #####
#################
# 24 23 22 21 20
# 19 18 17 16 15
# 14 13 12 11 10
#  9  8  7  6  5
#  4  3  2  1  0
#################

seenBoards = []
solution = []
finalMoves = []
flag = False
# Total number of moves available throughout the board
moves = [] * 36


def initialize(initial):
    try:
        initial_position = initial  # sys.argv[x]  # input from user from 0 to 14
        try:
            if not (int(initial_position) <= 14 and int(initial_position) >= 0):
                raise Exception("enter arguments between following array")

        except Exception as error:
            print("#" * 60)
            print("Error:", repr(error))
            print("#" * 60)
            count = 0
            for i in range(0, 5, 1):
                for k in range(0, 5 - i, 1):
                    if (count < 10):
                        print("  ", end='', sep='', flush=True)
                for j in range(0, i + 1, 1):
                    print("(", count, ")", end='', sep='', flush=True)
                    print(" ", end='', sep='', flush=True)
                    count += 1
                print()
            sys.exit()
    except IndexError:
        print("#" * 84)
        print("Invalid Syntax: try - python3 HiqOld.py 12 {i.e. argument in between 0 to 14 inclusive}")
        print("#" * 84)
        sys.exit()

    print("# You have entered --> ", initial_position)
    print("-" * 25)
    print("     initial Board")
    print("-" * 25)

    boardBits = ""
    boardOppBits = ""
    counter = 0

    # printing the initial position of board
    for i in range(0, 5, 1):
        print("  ", end='', sep='', flush=True)
        for k in range(0, 4 - i, 1):
            print("  ", end='', sep='', flush=True)

            boardOppBits += "0"
        for j in range(0, 5, 1):
            if (j <= i):
                if (counter == int(initial_position)):
                    print("o   ", end='', sep='', flush=True)
                    boardOppBits += "1"
                else:
                    print("x   ", end='', sep='', flush=True)
                    boardOppBits += "0"
                counter += 1
        print()

    print("-" * 25)
    print("         Starting ")
    print("-" * 25)
    # print(boardOppBits)
    boardBits = int(boardOppBits, 2) ^ 1154559
    # print(boardBits)

    # initial board (one marble free in given location by user )
    INITIAL_BOARD = boardBits
    # INITIAL_BOARD = 105983
    # print(INITIAL_BOARD)

    # goal board (one marble in given location by user)
    if (int(initial_position) == 4):
        GOAL_BOARD = 4
    elif (int(initial_position) == 8):
        GOAL_BOARD = 4096
    elif (int(initial_position) == 7):
        GOAL_BOARD = 1024
    else:
        GOAL_BOARD = int(boardOppBits, 2)
    # GOAL_BOARD = 1048576
    # print(GOAL_BOARD)

    # board that contains a ball in every available slot, i.e. GOAL_BOARD | INITIAL_BOARD

    #VALID_BOARD_CELLS = GOAL_BOARD | INITIAL_BOARD
    # print(VALID_BOARD_CELLS)
    VALID_BOARD_CELLS = 1154559

    return INITIAL_BOARD, GOAL_BOARD, VALID_BOARD_CELLS


# print the board
def printBoard(board, VALID_BOARD_CELLS):
    # loop over all cells (the board is 5 x 5)
    for i in range(24, -1, -1):
        validCell = ((1 << i) & VALID_BOARD_CELLS) != 0;
        if (validCell):
            if (((1 << i) & board) != 0):
                print("  x ", end='', sep='', flush=True)
            else:
                print("  o ", end='', sep='', flush=True)
        else:
            print("  ", end='', sep='', flush=True)
        if (i % 5 == 0):
            print()


# create the two possible moves for the three added pegs
def createMoves(bit1, bit2, bit3, moves):
    possiblemoves = []
    possiblemoves.append(1 << bit1)
    possiblemoves.append((1 << bit2) | (1 << bit3))
    possiblemoves.append((1 << bit1) | (1 << bit2) | (1 << bit3))

    moves.append(possiblemoves)

    possiblemoves = []
    possiblemoves.append(1 << bit3)
    possiblemoves.append((1 << bit2) | (1 << bit1))
    possiblemoves.append((1 << bit1) | (1 << bit2) | (1 << bit3))

    moves.append(possiblemoves)


# do the calculation recursively by starting from
# the "GOAL_BOARD" and doing moves in reverse

def find_solution(board, initialboard):
    # for all possible moves
    for move in moves:
        # check if the move is valid
        if ((move[1] & board) == 0 and (move[0] & board) != 0):
            # calculate the board after this move was applied
            newBoard = board ^ move[2]
            # only continue processing if we have not seen this board before
            contains = False
            for x in range(0, len(seenBoards), 1):
                if (int(seenBoards[x]) == int(newBoard)):
                    contains = True
            if not (contains):
                seenBoards.append(newBoard)
                # check if the initial board is reached
                if (newBoard == initialboard or find_solution(newBoard, initialboard)):
                    finalMoves.append(move)
                    solution.append(board)
                    return True
    return False


def convertMove(move):
    Newmove = str(bin(move[2])[2:])
    # [print("move: " + bin(x)[2:]) for x in move]
    count = 0
    tof = []

    count1 = 5
    if (len(Newmove) >= 5):
        for x in range(len(Newmove) - 1, -1, -5):
            for y in range(x, x - count1, -1):
                if (int(Newmove[y]) == 0):
                    count += 1
                if (int(Newmove[y]) == 1):
                    count += 1
                    tof.append(15 - count)
            count1 -= 1
    else:
        for y in range(len(Newmove) - 1, -1, -1):
            if (int(Newmove[y]) == 0):
                count += 1
            if (int(Newmove[y]) == 1):
                count += 1
                tof.append(15 - count)
    print("--" * 14)
    print("[from,over,to]=", tof)
    print("--" * 14)
    return tof


def main(initialPos):
    INITIAL_BOARD, GOAL_BOARD, VALID_BOARD_CELLS = initialize(initialPos)
    while True:
        moves.clear()
        seenBoards.clear()
        solution.clear()
        finalMoves.clear()
        solution.append(INITIAL_BOARD)
        # holds all starting positions in west-east direction
        startsX = {4, 3, 2, 8, 7, 12}
        for x in startsX:
            createMoves(x, x - 1, x - 2, moves)

        # holds all starting positions in north-south direction
        startsY = {20, 15, 10, 16, 11, 12}
        for y in startsY:
            createMoves(y, y - 5, y - 10, moves)

        startsZ = {4, 8, 12, 3, 7, 2}
        for z in startsZ:
            createMoves(z, z + 4, z + 8, moves)
        shuffle(moves)
        startTime = time.time()

        find_solution(GOAL_BOARD, INITIAL_BOARD)

        i = 0
        for step in solution:
            printBoard(step, VALID_BOARD_CELLS)
            if (i < len(finalMoves)):
                # print(finalMoves[i])
                convertMove(finalMoves[i])
            i += 1
        print("-" * 26)
        print("Total No. of moves: ", len(finalMoves))
        print("-" * 44)
        print("Completed in: ", (time.time() - startTime), "seconds")
        print("-" * 44)

        if (input("Find more solutions?(yes/no): ") == "no"):
            sys.exit()
            print("+" * 45)
            print("################# Thank You #################")
            print("+" * 45)


InitialPosition = input("Enter initial position:")
if not InitialPosition == None:
    main(InitialPosition)

File: Hiq.java
import java.util.*;

/**
 * Created by Vishal Kulkarni on 11/29/16.
 *    VVK27 | CS635 | vvk27@njit.edu
 */

/*
##### BOARD ####
# 24 23 22 21 20
# 19 18 17 16 15
# 14 13 12 11 10
#  9  8  7  6  5
#  4  3  2  1  0
 */

public class Hiq {

    // list of seen boards - this is used to prevent rechecking of paths
    private static final HashSet<Long> seenBoards = new HashSet<Long>();

    // list of solution boards in ascending order - filled in once the solution is found
    private static final ArrayList<Long> solution = new ArrayList<Long>();

    private static final ArrayList<long[]> finalMoves = new ArrayList<long[]>();

    public static void setGoalBoard(long goalBoard) {
        GOAL_BOARD = goalBoard;
    }

    public static void setInitialBoard(long initialBoard) {
        INITIAL_BOARD = initialBoard;
    }

    // goal board (one marble in center)
    private static long GOAL_BOARD = 1048576;


    // initial board (one marble free in center)
    private static long INITIAL_BOARD = 105983;


    // board that contains a ball in every available slot, i.e. GOAL_BOARD | INITIAL_BOARD
    private static final long VALID_BOARD_CELLS = 1154559;


    static void initialize(int initial_position) {
        int boardBits = 0;
        String boardOppBits = "";
        int counter = 0;

        // printing the initial position of board
        for (int i = 0; i < 5; i++) {
            System.out.print("  ");
            for (int k = 0; k < 4 - i; k++) {
                System.out.print("  ");
                boardOppBits += "0";
            }

            for (int j = 0; j < 5; j++) {
                if (j <= i) {
                    if (counter == initial_position) {
                        System.out.print("o   ");
                        boardOppBits += "1";
                    } else {
                        System.out.print("x   ");
                        boardOppBits += "0";
                    }

                    counter += 1;
                }

            }
            System.out.println();

        }


        System.out.println("----------------------");
        System.out.println("       Starting ");
        System.out.println("----------------------");
        //print(boardOppBits)
        boardBits = Integer.parseInt(boardOppBits, 2) ^ 1154559;
        //print(boardBits)

        //initial board (one marble free in given location by user )
        setInitialBoard(boardBits);
        //INITIAL_BOARD = 105983
        //print(INITIAL_BOARD)

        //goal board (one marble in given location by user)
        if ((initial_position) == 4)
            setGoalBoard(4);
        else if ((initial_position) == 8)
            setGoalBoard(4096);
        else if(initial_position == 7)
            setGoalBoard(1024);
        else
            setGoalBoard(Integer.parseInt(boardOppBits, 2));
        //GOAL_BOARD = 1048576
        //print(GOAL_BOARD)


    }


    // holds all 36 moves that are possible
    // the inner array is structures as following:
    // - first entry holds the peg that is added by the move
    // - second entry holds the two pegs that are removed by the move
    // - third entry holds all three involved pegs
    private static final long[][] moves = new long[36][];

    // -------

    // print the board
    private static void printBoard(long board) {
        // loop over all cells (the board is 5 x 5)
        for (int i = 24; i > -1; i--) {
            boolean validCell = ((1L << i) & VALID_BOARD_CELLS) != 0L;
            System.out.print(validCell ? (((1L << i) & board) != 0L ? "x   " : "o   ") : "  ");
            if (i % 5 == 0) System.out.println();
        }
        //System.out.println("--------------------");
    }

    // create the two possible moves for the three added pegs
    // (this function assumes that the pegs are in one continuous line)
    private static void createMoves(int bit1, int bit2, int bit3, ArrayList<long[]> moves) {
        moves.add(new long[]{(1L << bit1), (1L << bit2) | (1L << bit3), (1L << bit1) | (1L << bit2) | (1L << bit3)});
        moves.add(new long[]{(1L << bit3), (1L << bit2) | (1L << bit1), (1L << bit1) | (1L << bit2) | (1L << bit3)});

        //System.out.println("moves" + moves);
    }

    // do the calculation recursively by starting from
    // the "GOAL_BOARD" and doing moves in reverse
    private static boolean search(long board) {
        //int here =0;
        // for all possible moves
        for (long[] move : moves) {
            //System.out.println(here++);
            // check if the move is valid
            if ((move[1] & board) == 0L && (move[0] & board) != 0L) {
                // calculate the board after this move was applied
                long newBoard = board ^ move[2];
                // only continue processing if we have not seen this board before
                if (!seenBoards.contains(newBoard)) {
                    seenBoards.add(newBoard);
                    // check if the initial board is reached
                    if (newBoard == INITIAL_BOARD || search(newBoard)) {
                        solution.add(board);
                        finalMoves.add(move);
                        return true;
                    }
                }
            }
        }

        return false;
    }


    private static void convertMove(long[] moves) {
        String Newmove = Long.toBinaryString(moves[2]);
        int count = 0;
        ArrayList<Integer> tof = new ArrayList<>();


        int count1 = 5;
        if ((Newmove).length() >= 5) {
            for (int x = ((Newmove).length() - 1); x > -1; x -= 5) {
                for (int y = x; y > (x - count1); y--) {
                    if (y > -1 && (Newmove.charAt(y)) == '0')
                        count += 1;
                    if (y > -1 && (Newmove.charAt(y)) == '1') {
                        count += 1;
                        tof.add(15 - count);
                    }
                }
                count1--;
            }
        } else {

            for (int y = (Newmove).length() - 1; y > -1; y--) {
                if (y > -1 && (Newmove.charAt(y)) == '0')
                    count += 1;
                if (y > -1 && (Newmove.charAt(y)) == '1') {
                    count += 1;
                    tof.add(15 - count);
                }
            }

        }

        System.out.println("------------------------");
        System.out.print("[from,over,to]= ");
        for (int i = 0; i < 3; i++) {
            System.out.print(tof.get(i) + " ");

        }
        System.out.println();
        System.out.println("------------------------");
    }

    // the main method
    public static void main(String[] args) {
        // to measure the overall runtime of the program
        long time = System.currentTimeMillis();

        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter initial position of your peg: ");
        int initialPos = scanner.nextInt();
        if (!(initialPos >= 0 && initialPos <= 14)) {
            System.out.println("invalid output please enter number between 0 to 14");
            System.exit(0);
        }
        initialize(initialPos);


        // add starting board (as this board is not added by the recursive function)
        solution.add(INITIAL_BOARD);

        // generate all possible moves
        //ArrayList<long[]> moves = new ArrayList<long[]>();
        // holds all starting positions in west-east direction
        //int[] startsX = new int[]{24, 23, 22, 19, 18, 14};


        // fill in the global moves variable that is used by the solver
        //moves.toArray(Hiq.moves);

        boolean run = true;
        while (run) {
            seenBoards.clear();
            //moves.clear();
            ArrayList<long[]> moves = new ArrayList<long[]>();
            int[] startsX = new int[]{4, 3, 2, 8, 7, 12};
            for (int x : startsX) {
                createMoves(x, x - 1, x - 2, moves);
            }
            // holds all starting positions in north-south direction
            //int[] startsY = new int[]{24, 23, 22, 19, 18, 14};
            int[] startsY = new int[]{20, 15, 10, 16, 11, 12};
            for (int y : startsY) {
                createMoves(y, y - 5, y - 10, moves);
            }
            // holds all starting positions in north-south direction
            //int[] startsZ = new int[]{4, 8, 12, 9, 13, 14};
            int[] startsZ = new int[]{4, 8, 12, 3, 7, 2};
            for (int z : startsZ) {
                createMoves(z, z + 4, z + 8, moves);
            }

            Collections.shuffle(moves);
            moves.toArray(Hiq.moves);
            // start recursively search for the initial board from the goal (reverse direction!)
            search(GOAL_BOARD);


            // the found solution
            int i = 0;
            for (long step : solution) {
                //System.out.println("solutions"+solution);
                printBoard(step);

                if (i < (finalMoves).size()) {
                    convertMove(finalMoves.get(i));
                    i += 1;
                }
            }
            System.out.println("------------------------");
            System.out.println("Completed in " + (System.currentTimeMillis() - time) + " ms.");
            System.out.println("------------------------");

            scanner = new Scanner(System.in);
            System.out.print("Find more solutions?(yes/no): ");
            String Answer = scanner.next();

            if (Answer.equals("no"))
                run = false;
        }


    }
}
